allennlp/models/model.py

```diff
@@ -147,7 +182,18 @@ def get_metrics(self, reset: bool = False) -> Dict[str, float]:
     @classmethod
     def from_params(cls, vocab: Vocabulary, params: Params) -> 'Model':
         choice = params.pop_choice("type", cls.list_available())
-        return cls.by_name(choice).from_params(vocab, params)
+        regularizer_params = params.pop("regularizer", None)
+        initializer_params = params.pop("initializer", None)
+
+        model = cls.by_name(choice).from_params(vocab, params)
+
+        if regularizer_params is not None:
```

Comment:
I'm very unsure that this is a good solution. This makes building a model `from_params` substantially different to building one using the constructor. As soon as this happens once, we will keep being tempted to add things that are useful for our specific training regime to the `from_params` constuctor, but make it effectively impossible to use a model in allennlp directly. 

I would vote every time for adding 4 lines to a new model constructor (two parameters, one to call super and one to initialize the model), which is explicit in how it works and explicit in how it breaks if you don't use it correctly. E.g we could log a warning in the base model constructor if you don't pass an initializer.

Reply:
yeah, I think I am coming around to this point of view

Reply:
I don't have strong opinions on this at this point; I can see pros and cons of doing it both ways.  I'd like to avoid required boilerplate, though.  So, here's a proposed solution that hopefully will make Mark happy:

Because of how regularization interacts with the `Trainer`, the superclass _has_ to know about it, so it needs to get it somehow.  I'm fine with passing it explicitly into the constructor, and having it be `None` by default, because it's not typical for models to use regularization.  A subclass can just totally ignore regularization, and that's fine, so there's no required extra boilerplate.  You just need to add some code if you want to try out regularization.  Not a huge deal.

Also, we should try to make the default initialization of everything we use reasonable, so that it's not typical to need an initializer.  Then most models can just ignore initialization also, and not have anything extra needed there.  I know I recommended putting the initialization in the super class, but seeing that it requires you to actually _run_ the initializer outside of the constructor makes me think that was a bad suggestion.  Here again we can just put this back how it was, with the model adding the ability to do initialization if it wants.  It's not that much boilerplate, and it's not required for a new model.